File: Source/Game/Systems/ReconnectSystem.cpp
#include "ReconnectSystem.h"
#include "../NetworkInstance.h"

ReconnectSystem::ReconnectSystem()
{
	m_timer = 0;
}
ReconnectSystem::~ReconnectSystem()
{
}

void ReconnectSystem::Initialize()
{
	/*	Rendersystem wants Network	*/
	AddComponentTypeToFilter("Username", ECSL::FilterType::Mandatory);
	AddComponentTypeToFilter("NetConnection", ECSL::FilterType::Mandatory);

	printf("ReconnectSystem initialized!\n");


	std::function<void(Network::NetConnection, const char*)> hook = std::bind(&ReconnectSystem::OnUserDisconnected, this, std::placeholders::_1, std::placeholders::_2);
	NetworkInstance::GetServer()->SetOnPlayerDisconnected(hook);

	//NetworkInstance::GetServer()->SetOnPlayerTimedOut(hook);
}

void ReconnectSystem::Update(float _dt)
{
	
}

void ReconnectSystem::OnEntityAdded(unsigned int _entityId)
{
	bool match = false;
	unsigned int matchId = 0;

	char* username		= GetComponent(_entityId, "Username", "Name");
	char* ipAddress		= GetComponent(_entityId, "NetConnection", "IpAddress");
	unsigned int port	= *(int*)GetComponent(_entityId, "NetConnection", "Port");


	auto entities = *GetEntities();

	for (int i = 0; i < entities.size(); ++i)
	{
		if (_entityId == entities[i])
			continue;

		if (strcmp(username, GetComponent(entities[i], "Username", 0)) == 0)
		{
			match = true;
			matchId = entities[i];
			break;
		}

	}

	if (match) // Found an already existing user
	{
		bool setActive = true;

		if (*(bool*)GetComponent(matchId, "NetConnection", "Active")) // If the matching user is still connected kick the new user
		{
			char* oldIp = GetComponent(matchId, "NetConnection", "IpAddress");
			unsigned int oldPort = *(int*)GetComponent(matchId, "NetConnection", "Port");

			NetworkInstance::GetServer()->Kick(Network::NetConnection(oldIp, oldPort), "Connected somewhere else");
		}

		SetComponent(matchId, "NetConnection", "IpAddress", ipAddress);
		SetComponent(matchId, "NetConnection", "Port", &port);
		SetComponent(matchId, "NetConnection", "Active", &setActive);

		KillEntity(_entityId);

	}
	else
	{
		bool setActive = true;
		SetComponent(_entityId, "NetConnection", "Active", &setActive);
	}

}

void ReconnectSystem::OnEntityRemoved(unsigned int _entityId)
{

}

void ReconnectSystem::OnUserDisconnected(Network::NetConnection _nc, const char* _message)
{
	const char* ipAddress = _nc.GetIpAddress();
	unsigned int port = _nc.GetPort();

	auto entities = *GetEntities();

	for (int i = 0; i < entities.size(); ++i)
	{
		if (strcmp(ipAddress, GetComponent(entities[i], "NetConnection", "IpAddress")) == 0
			&& port == *(int*)GetComponent(entities[i], "NetConnection", "Port"))
		{
			bool setActive = false;
			SetComponent(entities[i], "NetConnection", "Active", &setActive);
		}
	}

}

File: Externals/content/scripting/storaspel/systems/reconnectsystem.lua
ReconnectSystem = System()

ReconnectSystem.Update = function(self, dt)

end

ReconnectSystem.OnUsername = function(id, ipAddress, port)
	Console.Print("ReconnectSystem.OnUsername");
	
	local name = "user" -- ipAddress .. port;
	local active = false;
	
	Console.Print("Name: " .. name);
	Console.Print("Id: " .. id);
	Console.Print("IpAddress: " .. ipAddress);
	Console.Print("Port: " .. tostring(port));
	
	local eId = world:CreateNewEntity("User");
	world:SetComponent(eId, "Username", "Name", name);
	world:SetComponent(eId, "NetConnection", "IpAddress", ipAddress);
	world:SetComponent(eId, "NetConnection", "Port", port);
	world:SetComponent(eId, "NetConnection", "Active", active);

end

ReconnectSystem.Initialize = function(self)
	self:SetName("Reconnect System")
	self:InitializeNetworkEvents()
	
	self:AddComponentTypeToFilter("Username", FilterType.Mandatory)
	self:AddComponentTypeToFilter("NetConnection", FilterType.Mandatory)
	
	Net.Receive("Username", ReconnectSystem.OnUsername);
	
	print("ReconnectSystem initialized!")
	
end
ReconnectSystem.OnEntityAdded = function(self, entityId)
	Console.Print("ReconnectSystem.OnEntityAdded");

	local match = false;
	local matchId = 0;
	
	Console.Print("Entityid: " .. entityId);
	local username 	= self:GetComponent(entityId, "Username", "Name");
	local ipAddress = self:GetComponent(entityId, "NetConnection", "IpAddress");
	local port 		= self:GetComponent(entityId, "NetConnection", "Port");
	
	
	local entites = self:GetEntities();
	for i = 1, #entites do
	
		Console.Print("Entityid2: " .. entites[i]);
	
		if entityId == math.floor(entites[i]) then
			Console.Print("id1: " .. entityId);
			Console.Print("id2: " .. entites[i]);
		else
			local uname = self:GetComponent(entites[i], "Username", "Name");
			Console.Print("Uname: " .. uname:GetString());
			
			if username:GetString() == uname:GetString() then
				match = true;
				matchId = math.floor(entites[i]);
				break;
			end
			
		end
		
	end

	if match then
		Console.Print("Match!")
		
		local setActive = true;
		local isActive = self:GetComponent(matchId, "NetConnection", "Active");
		
		if isActive:GetBool() then
			Console.Print("isActive");
			local oldIp 	= self:GetComponent(matchId, "NetConnection", "IpAddress");
			local oldPort	= self:GetComponent(matchId, "NetConnection", "Port");
			local reason	= "ReconnectSystem kicked you.";
			
			Server.Kick(oldIp:GetString(), oldPort:GetInt(), reason);
			
		end
		
		world:SetComponent(matchId, "NetConnection", "IpAddress", ipAddress:GetString());
		world:SetComponent(matchId, "NetConnection", "Port", port:GetInt());
		world:SetComponent(matchId, "NetConnection", "Active", setActive);
		
		world:KillEntity(entityId);
	else
		local setActive = true;
		world:SetComponent(entityId, "NetConnection", "Active", setActive);
	end
	
end
ReconnectSystem.OnEntityRemoved = function(self, entityId)

end